### PR TITLE
feat(j2cl): add search sidecar slice

### LINKED ISSUE #901: Migrate the search results panel as the first J2CL UI vertical slice
## Goal
Use the search/inbox results panel as the first real J2CL UI slice after the sidecar scaffold and transport rewrite exist.

## Why this slice
The current search stack is still very GWT-specific, but it is much smaller and safer than a full app-shell or editor migration. It is also explicitly called out in the J2CL plan as the best first UI vertical slice.

Current live search/UI seams include:
- `wave/src/main/java/org/waveprotocol/box/webclient/search/SearchPanelWidget.java`
- `wave/src/main/java/org/waveprotocol/box/webclient/search/SearchWidget.java`
- `wave/src/main/java/org/waveprotocol/box/webclient/search/DigestDomImpl.java`
- `wave/src/main/java/org/waveprotocol/box/webclient/search/SearchPresenter.java`
- `wave/src/main/java/org/waveprotocol/box/webclient/search/SearchPanelResourceLoader.java`
- `gen/messages/org/waveprotocol/box/search/jso/**`

The slice still depends on UiBinder, `ClientBundle`, `CssResource`, synchronous `StyleInjector`, DOM observers, and JSO search messages.

## Source docs
- `docs/superpowers/plans/j2cl-full-migration-plan.md` sections 2.3 and 8
- `docs/j2cl-gwt3-decision-memo.md`

## Scope
- preserve the existing presenter / query logic where possible
- replace the widget/view layer for the search results panel with explicit DOM construction on the J2CL side
- use static CSS or build-time CSS extraction instead of `ClientBundle` / `CssResource`
- consume the J2CL-friendly search codecs from the transport/codegen issue instead of `SearchRequestJsoImpl` / `SearchResponseJsoImpl`
- keep the slice narrow enough to validate dual-run rendering before broader UI migration

## Non-goals
- no full app-shell migration
- no editor migration
- saved-search editor popups can stay on the legacy path if they are not required for the first result-panel slice

## Acceptance Criteria
- the migrated search slice can execute a query and render digests from the J2CL side
- unread / mention / task badges and result-selection behavior still work in the migrated slice
- the first migrated search view no longer depends on UiBinder, `ClientBundle`, or JSO search message types
- local verification proves the J2CL slice can coexist with the legacy GWT bundle during the transition

### PR DESCRIPTION
**Pull request by vega113**
Closes #901

## Summary
- add the first real J2CL search results slice under `/j2cl-search/index.html`
- keep the legacy root app and legacy search path intact on `/`
- render the sidecar search panel with explicit DOM and static CSS instead of UiBinder/ClientBundle/CssResource
- consume the sidecar-safe search transport/codecs path and prove dual-run rendering against the same local server

## What changed
- extended the sidecar sandbox entrypoint to mount a real search/results panel
- added sidecar search controller/view/gateway/result projection classes under `j2cl/src/main/java/org/waveprotocol/box/j2cl/search/`
- added J2CL tests for query defaulting, result projection, show-more paging, and digest selection wiring
- added the issue-specific plan doc and changelog fragment for the search slice lane

## Verification
- `sbt -batch generatePstMessages "testOnly org.waveprotocol.pst.PstCodegenContractTest"`
  - passed
- `sbt -batch j2clSearchBuild j2clSearchTest`
  - passed
- `python3 scripts/assemble-changelog.py`
  - regenerated `wave/config/changelog.json`
- `python3 scripts/validate-changelog.py --changelog wave/config/changelog.json`
  - passed
- `sbt -batch compileGwt Universal/stage`
  - passed
- `bash scripts/worktree-boot.sh --port 9900`
  - passed
- `PORT=9900 JAVA_OPTS='-Djava.util.logging.config.file=/Users/vega/devroot/worktrees/issue-901-j2cl-search-slice/wave/config/wiab-logging.conf -Djava.security.auth.login.config=/Users/vega/devroot/worktrees/issue-901-j2cl-search-slice/wave/config/jaas.config' bash scripts/wave-smoke.sh start`
- `PORT=9900 bash scripts/wave-smoke.sh check`
  - `ROOT_STATUS=200`
  - `HEALTH_STATUS=200`
  - `WEBCLIENT_STATUS=200`
- route checks:
  - `/` -> `200 OK`
  - `/webclient/webclient.nocache.js` -> `200 OK`
  - `/j2cl-search/index.html` -> `200 OK`
- browser verification in one authenticated session:
  - `/` still loaded the legacy signed-in GWT app
  - `/j2cl-search/index.html` rendered the J2CL slice
  - default `in:inbox` rendered results
  - `with:@` rerendered to an explicit empty state without leaving the route
  - after seeding additional inbox waves for the same test account, the sidecar showed `30 of 54 waves · 30 unread`, exposed `Show more waves`, then expanded to `54 waves · 54 unread`
- `PORT=9900 bash scripts/wave-smoke.sh stop`
  - passed

## Notes
This PR intentionally does not cut over the root app shell or editor. The legacy root path remains the runtime source of truth on `/`; the J2CL search slice is isolated to `/j2cl-search/index.html`.

## Traceability
- Worktree: `/Users/vega/devroot/worktrees/issue-901-j2cl-search-slice`
- Plan: `docs/superpowers/plans/2026-04-19-issue-901-j2cl-search-slice.md`
- Local verification record: `journal/local-verification/2026-04-19-issue-901-j2cl-search-slice.md`

## Summary by CodeRabbit

* **New Features**
  * Isolated J2CL search-results sidecar served separately with query submission, live results, pagination/show-more, wave counts, unread/pinned indicators, selectable digests, and responsive mobile layout.

* **Documentation**
  * Added implementation plan and changelog entry for the J2CL search vertical slice and verification steps.

* **Tests**
  * New unit tests covering controller behavior, result projection, and sandbox parsing edge cases.

* **Style**
  * Updated sidecar CSS for search UI, loading states, and responsiveness.

* **Bug Fixes**
  * Hardened query and cookie decoding; improved transport/error reporting.

### CODE REVIEW

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        J2clDigestView digestView =
+            new J2clDigestView(
+                item,
+                waveId -> {
+                  if (listener != null) {
+                    listener.onDigestSelected(waveId);
+                  }
+                });
+        digestViews.put(item.getWaveId(), digestView);
+        digestList.appendChild(digestView.element());
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:177]: Already handled.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    showMoreButton.hidden = !model.isShowMoreVisible();
+  }
+
+  @Override
+  public void setSelectedWaveId(String waveId) {
+    for (Map.Entry<String, J2clDigestView> entry : digestViews.entrySet()) {
+      entry.getValue().setSelected(entry.getKey().equals(waveId));
+    }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:188]: Already handled: the comparison uses `entry.getKey() != null && entry.getKey().equals(waveId)`, and null keys never enter the map since `render()` skips them.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    view.setStatus("Bootstrapping the root session.", false);
+    gateway.fetchRootSessionBootstrap(
+        bootstrap -> {
+          view.setSessionSummary(bootstrap.getAddress());
+          requestSearch();
+        },
+        error -> {
+          view.setSessionSummary("Using the current browser session.");
+          view.setStatus(
+              "Root bootstrap lookup failed; continuing with the current browser session.",
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java]: Already handled: error callback uses `error=false` and the text already reads "Continuing with the current browser session." (neutral severity).

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+  public interface WaveSelectionHandler {
+    void onWaveSelected(String waveId);
+  }
+
+  private final SearchGateway gateway;
+  private final View view;
+  private final WaveSelectionHandler selectionHandler;
+  private final double viewportWidth;
+  private final int pageIncrement;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java:51]: Already handled: `viewportWidth` is used directly in the constructor to compute `pageIncrement` and is never stored as a field.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
+package org.waveprotocol.box.j2cl.search;
+
+import java.util.ArrayList;
+import java.util.Collections;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java]: Not applicable: `java.util.Collections` is not imported in `J2clSearchResultProjector` — it is imported and used in `SandboxEntryPoint` for `Collections.singletonList`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java [hunk 6]
```diff
@@ -101,6 +116,25 @@ private static String normalizeMode(String requestedMode) {
...
+  static String readRequestedQuery(String search) {
+    if (search == null || search.isEmpty()) {
+      return DEFAULT_QUERY;
+    }
+    String trimmed = search.charAt(0) == '?' ? search.substring(1) : search;
+    String[] parts = trimmed.split("&");
+    for (String part : parts) {
+      if (part.startsWith("q=")) {
+        String value = part.substring(2);
+        return value.isEmpty() ? DEFAULT_QUERY : decodeUriComponent(value);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java]: Fixed in de2872b: wrapped `decodeUriComponent` in try/catch in `readRequestedQuery`, falling back to `DEFAULT_QUERY` on URIError.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java [hunk 7]
```diff
@@ -101,6 +116,25 @@ private static String normalizeMode(String requestedMode) {
...
+      }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java:140]: Fixed in de2872b: same try/catch guard.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        });
+  }
+
+  @Override
+  public void onQuerySubmitted(String query) {
+    currentQuery = J2clSearchResultProjector.normalizeQuery(query);
+    currentPageSize = pageIncrement;
+    selectedWaveId = null;
+    view.setSelectedWaveId(null);
+    requestSearch();
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java:97]: Fixed in de2872b: added `clearSelection()` helper that sets `selectedWaveId = null`, calls `view.setSelectedWaveId(null)`, and notifies `selectionHandler`. All three clear-selection sites now use it.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    HTMLFormElement form = (HTMLFormElement) DomGlobal.document.createElement("form");
+    form.className = "sidecar-search-toolbar";
+    card.appendChild(form);
+
+    queryInput = (HTMLInputElement) DomGlobal.document.createElement("input");
+    queryInput.className = "sidecar-search-input";
+    queryInput.type = "search";
+    queryInput.placeholder = "Search waves";
+    queryInput.autocomplete = "off";
+    form.appendChild(queryInput);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:68]: Fixed in de2872b: added `queryInput.setAttribute("aria-label", "Search waves")`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      }
+    }
+
+    showMoreButton.hidden = !model.isShowMoreVisible();
+  }
+
+  @Override
+  public void setSelectedWaveId(String waveId) {
+    for (Map.Entry<String, J2clDigestView> entry : digestViews.entrySet()) {
+      entry.getValue().setSelected(entry.getKey().equals(waveId));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java]: Already handled: `render()` skips null wave IDs before inserting into the map, and `setSelectedWaveId` uses `entry.getKey() != null &&` guard.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    } else {
+      waveCountText = total + " waves";
+    }
+    if (unreadWaveCount > 0) {
+      waveCountText += " \u00b7 " + unreadWaveCount + " unread";
+    }
+
+    boolean showMoreVisible =
+        totalKnown ? loaded < total : (requestedSize > 0 && loaded >= requestedSize);
+    return new J2clSearchResultModel(items, waveCountText, showMoreVisible, "");
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java:75]: Fixed in 51d4ac5: added `if (digest.getWaveId() == null) { continue; }` guard before the unread/item accumulation, and a post-loop `items.isEmpty()` check that returns the empty model — matching `J2clSearchPanelView.render` behavior.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:177]: Handled in `be54e38b2`. `J2clSearchPanelView.render(...)` now skips digests with a null `waveId`, so they never enter `digestViews` or the rendered list.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:188]: Handled in `be54e38b2`. `setSelectedWaveId(...)` now guards null keys before comparison, so selection refresh stays null-safe even if upstream data is incomplete.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java]: Handled in `de2872b52` and covered in `423d710bd`. `readRequestedQuery(...)` now catches malformed `decodeUriComponent` input and falls back to `DEFAULT_QUERY`; `SandboxBuildSmokeTest` covers the malformed `%` query case.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java:140]: Handled in `de2872b52` and covered in `423d710bd`. Both decode sites are now guarded: malformed query values fall back to `DEFAULT_QUERY`, and malformed cookie values return `null` instead of crashing startup. Added smoke coverage for both cases.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelController.java:97]: Handled in `de2872b52` and covered in `423d710bd`. Selection clears now flow through `clearSelection()` to both the view and `WaveSelectionHandler`, and the controller tests cover submit, missing-result, and error clear paths.

**Comment by vega113** [comment]
Fixed in 9c47be3: extracted `safeDecodeUriComponent(value, defaultValue)` helper that catches `RuntimeException | LinkageError` and replaced both `decodeUriComponent` call sites (`readRequestedQuery` and `readCookie`) to use it.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java [hunk 12]
```diff
@@ -101,6 +116,32 @@ private static String normalizeMode(String requestedMode) {
...
+    String trimmed = search.charAt(0) == '?' ? search.substring(1) : search;
+    String[] parts = trimmed.split("&");
+    for (String part : parts) {
+      if (part.startsWith("q=")) {
+        String value = part.substring(2);
+        if (value.isEmpty()) {
+          return DEFAULT_QUERY;
+        }
+        try {
+          return decodeUriComponentSafe(value);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java]: Fixed in 79ec4a7: added `.replace("+", " ")` before calling `safeDecodeUriComponent` in `readRequestedQuery`, so form-encoded bookmarked URLs like `?q=in%3Ainbox+unread%3Atrue` now decode correctly to `in:inbox unread:true`.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java]: Fixed in `81514adf1`: `readRequestedQuery(...)` now treats `+` as a space before URI decoding, and `SandboxBuildSmokeTest.encodedQueryTreatsPlusAsSpace` covers `?q=mentions%3Ame+unread%3Atrue`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+      waveCountText = loaded + " of " + total + " waves";
+    } else {
+      waveCountText = total + " waves";
+    }
+    if (unreadWaveCount > 0) {
+      waveCountText += " \u00b7 " + unreadWaveCount + " unread";
+    }
+
+    boolean showMoreVisible =
+        totalKnown ? loaded < total : (requestedSize > 0 && loaded >= requestedSize);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java]: Fixed in `81514adf1`: `J2clSearchResultProjector` now keeps `showMoreVisible` true whenever `totalResults` is unknown, matching `SearchPresenter.renderShowMore()`. Covered by `projectKeepsShowMoreVisibleWhenTotalIsUnknown`.

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java]: Fixed in 79ec4a7: changed `showMoreVisible` to `!totalKnown || loaded < total`, matching legacy `SearchPresenter` behaviour (line 1042: `total == UNKNOWN_SIZE || querySize < total`). The previous heuristic could prematurely hide the button on a short page when the backend doesn't report a total.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+                item,
+                waveId -> {
+                  if (listener != null) {
+                    listener.onDigestSelected(waveId);
+                  }
+                });
+        digestViews.put(item.getWaveId(), digestView);
+        digestList.appendChild(digestView.element());
+      }
+    }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:179]: Already handled in commit 77d6900ce: added a `hasRenderableDigest` flag in the `else` branch; if all digests are skipped (null waveId), the flag stays false and `emptyState` is shown with the standard empty message.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  public static J2clSearchResultModel project(SidecarSearchResponse response, int requestedSize) {
+    if (response == null || response.getDigests().isEmpty()) {
+      return J2clSearchResultModel.empty("No waves matched this query.");
+    }
+
+    List<J2clSearchDigestItem> items = new ArrayList<J2clSearchDigestItem>();
+    int unreadWaveCount = 0;
+    for (SidecarSearchResponse.Digest digest : response.getDigests()) {
+      if (digest.getWaveId() == null) {
+        continue;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchResultProjector.java:38]: Fixed in befd288eb: added `digest == null ||` guard before `digest.getWaveId() == null` check on line 37 of J2clSearchResultProjector.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java [hunk 16]
```diff
@@ -101,6 +116,32 @@ private static String normalizeMode(String requestedMode) {
...
+    for (String part : parts) {
+      if (part.startsWith("q=")) {
+        String value = part.substring(2);
+        if (value.isEmpty()) {
+          return DEFAULT_QUERY;
+        }
+        try {
+          return decodeUriComponent(value);
+        } catch (RuntimeException e) {
+          return DEFAULT_QUERY;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/sandbox/SandboxEntryPoint.java:138]: Already fixed in 9c47be37b: extracted `safeDecodeUriComponent(value, defaultValue)` helper (line 457) that catches `RuntimeException | LinkageError`; both call sites (`readRequestedQuery` and the cookie path) now route through that helper.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/search/J2clSearchPanelView.java:179]: Already handled: `render()` tracks a `hasRenderableDigest` flag and shows the empty-state panel (with the model's empty message) when all digests were skipped due to null wave IDs. No blank panel scenario.
